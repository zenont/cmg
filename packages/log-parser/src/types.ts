import type { SensorType } from '@365-widgets/core'
import type { BadLineFormatError } from './error'

export interface GenericSensorSpec {
  readonly threshold: number
}

export interface ThermometerSpec extends GenericSensorSpec {
  readonly deviation: number
}

export type ThermometerSpecConfig = {
  readonly [SensorType.Thermometer]: ThermometerSpec
}

export type HumiditySensorSpecConfig = {
  readonly [SensorType.Humidity]: GenericSensorSpec
}

export type CarbonMonoxideSensorSpecConfig = {
  readonly [SensorType.CarbonMonoxide]: GenericSensorSpec
}

export type SpecConfig =
  | ThermometerSpecConfig
  | HumiditySensorSpecConfig
  | CarbonMonoxideSensorSpecConfig

export type QualityLogResult = Record<string, string>

export interface ReferenceResult {
  /**
   * The control room was held at this constant temperature.
   */
  readonly temperature: number
  /**
   * The control room relative humidity.
   */
  readonly humidity: number
  /**
   * The control room carbon monoxide.
   */
  readonly ppm: number
}

export interface SensorLineResultPayload {
  readonly type: SensorType
  readonly name: string
}

export interface ReferenceLineResult {
  readonly type: 'reference'
  readonly payload: ReferenceResult | BadLineFormatError
}

export interface SensorLineResult {
  readonly type: 'sensor'
  readonly payload: SensorLineResultPayload | BadLineFormatError
}

export interface EntryLineResult {
  readonly type: 'entry'
  readonly payload: number | BadLineFormatError
}

export type LineResult =
  | ReferenceLineResult
  | SensorLineResult
  | EntryLineResult

export interface SensorRead {
  readonly sensorType: SensorType
  readonly max: number
  readonly min: number
  readonly sumValue: number
  readonly totalCount: number
}
