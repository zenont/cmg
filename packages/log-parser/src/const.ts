export const TEMPERATURE_ULTRA_PRECISE_THRESHOLD = 3
export const TEMPERATURE_VERY_PRECISE_THRESHOLD = 5
export const TEMPERATURE_DEGREE_THRESHOLD = 0.5
export const HUMIDITY_THRESHOLD = 1
export const MONOXIDE_PPM_THRESHOLD = 3
