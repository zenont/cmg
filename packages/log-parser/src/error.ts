export class BadLineFormatError extends Error {
  constructor(line: string, lineNumber: number) {
    const formattedLineNumber = Intl.NumberFormat().format(lineNumber)
    super(`Unexpected or invalid line at ${formattedLineNumber}: [${line}]`)
  }
}
