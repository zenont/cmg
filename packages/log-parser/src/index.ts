export * from './types'
export * from './io'
export * from './error'
export * from './util'
export * from './parsers'
export * from './validators'
export * from './evaluator'
