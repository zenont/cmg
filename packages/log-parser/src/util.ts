export const splitLine = (line: string) => line.split(' ').map((x) => x.trim())
