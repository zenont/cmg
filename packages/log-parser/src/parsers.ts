import { isSensorType } from '@365-widgets/core'
import { BadLineFormatError } from './error'
import type {
  ReferenceLineResult,
  SensorLineResult,
  EntryLineResult,
  LineResult,
} from './types'
import { splitLine } from './util'

export const parseReferenceLine = (
  line: string,
  lineNumber: number
): ReferenceLineResult | undefined => {
  const [lineType, ...rest] = splitLine(line)

  if (lineType !== 'reference') {
    return undefined
  }

  const parsedSpec = rest
    .map((x) => Number.parseFloat(x))
    .filter((x) => !Number.isNaN(x))

  const [temperature, humidity, ppm] = parsedSpec
  return {
    type: 'reference',
    payload:
      parsedSpec.length !== 3
        ? new BadLineFormatError(line, lineNumber)
        : {
            temperature,
            humidity,
            ppm,
          },
  }
}

export const parseSensorLine = (
  line: string,
  lineNumber: number
): SensorLineResult | undefined => {
  const [lineType, ...rest] = splitLine(line)

  if (!isSensorType(lineType)) {
    return undefined
  }

  return {
    type: 'sensor',
    payload:
      rest[0] == null || rest[0].trim() === ''
        ? new BadLineFormatError(line, lineNumber)
        : { type: lineType, name: rest[0] },
  }
}

export const parseEntryLine = (
  line: string,
  lineNumber: number
): EntryLineResult | undefined => {
  const [, strValue] = splitLine(line)
  const value = Number.parseFloat(strValue)

  return {
    type: 'entry',
    payload: Number.isNaN(value)
      ? new BadLineFormatError(line, lineNumber)
      : value,
  }
}

/**
 * Parses a line of text from the log file. Validates it in the process to handle malformed or out of spec log files.
 * @param line The log file line.
 * @param lineNumber The log file number.
 * @returns Returns a {@link LineResult} if parsed correctly, or *undefined* if a valid line could not be parsed.
 */
export const parseLine = (
  line: string,
  lineNumber: number
): LineResult | undefined => {
  if (line == null || line.trim() === '') {
    return undefined
  }

  const tuple =
    parseReferenceLine(line, lineNumber) ??
    parseSensorLine(line, lineNumber) ??
    parseEntryLine(line, lineNumber)

  return tuple
}
