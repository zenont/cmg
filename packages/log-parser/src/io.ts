import { Readable, Writable } from 'stream'
import { createInterface } from 'readline'
import type { LineResult } from './types'
import { parseLine } from './parsers'

export type YieldedLineTuple = [LineResult | undefined, number, string]

/**
 * Reads a line at a time from a file. Yields a tuple {@link YieldedLineTuple}
 * @param logContentsStr The file contents as a string.
 */
export async function* readLines(
  logContentsStr: string
): AsyncGenerator<YieldedLineTuple, void, unknown> {
  // we use streams to account for big files
  const readStream = Readable.from(logContentsStr)
  const writeStream = new Writable()
  const reader = createInterface({
    input: readStream,
    output: writeStream,
    terminal: false,
  })

  let lineNumber = 0
  for await (const line of reader) {
    lineNumber++
    const parsedLine = parseLine(line, lineNumber)
    yield [parsedLine, lineNumber, line]
  }
}
