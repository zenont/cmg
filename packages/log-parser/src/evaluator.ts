import { SensorType } from '@365-widgets/core'
import { readLines } from './io'
import { validateSensors } from './validators'
import type { ReferenceResult, QualityLogResult, SensorRead } from './types'

export async function evaluateLogFile(
  logContentsStr: string
): Promise<QualityLogResult> {
  let currReference: ReferenceResult | undefined = undefined
  let currSensorKey: string | undefined = undefined
  let currSensorType: SensorType | undefined = undefined
  const sensors = new Map<string, SensorRead>()
  for await (const [line] of readLines(logContentsStr)) {
    if (line == null) {
      continue
    }

    if (line.payload instanceof Error) {
      throw line.payload
    }

    if (line.type === 'reference') {
      // we scope the reference
      currReference = line.payload
    } else if (line.type === 'sensor') {
      // we scope the sensor
      currSensorKey = line.payload.name
      currSensorType = line.payload.type
    } else if (line.type === 'entry') {
      // we don't include entries if they are not scoped to a sensor
      if (
        currSensorKey == null ||
        currSensorType == null ||
        currReference == null
      ) {
        continue
      }

      const sensor = sensors.get(currSensorKey)
      const sensorType = sensor?.sensorType ?? currSensorType
      const sumValue = (sensor?.sumValue ?? 0) + line.payload
      const totalCount = (sensor?.totalCount ?? 0) + 1
      const min =
        sensor?.min != null ? Math.min(sensor.min, line.payload) : line.payload
      const max =
        sensor?.max != null ? Math.max(sensor.max, line.payload) : line.payload

      sensors.set(currSensorKey, {
        ...sensor,
        sensorType,
        min,
        max,
        sumValue,
        totalCount,
      })
    }
  }

  if (currReference == null) return {}
  return validateSensors(currReference, sensors)
}
