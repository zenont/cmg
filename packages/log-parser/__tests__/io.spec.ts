import path from 'path'
import fs from 'fs'
import { readLines } from '../src'

describe('log-parser', () => {
  describe('io', () => {
    it('reads the file lines', async () => {
      const filePath = path.join(__dirname, './', 'sample.log')
      const logContentsStr = fs.readFileSync(filePath, 'utf8')
      const arr = []

      for await (const line of readLines(logContentsStr)) {
        arr.push(line)
      }

      expect(arr).toMatchSnapshot()
    })
  })
})
