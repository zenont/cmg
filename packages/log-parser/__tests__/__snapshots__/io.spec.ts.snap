// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`log-parser io reads the file lines 1`] = `
Array [
  Array [
    Object {
      "payload": Object {
        "humidity": 45,
        "ppm": 6,
        "temperature": 70,
      },
      "type": "reference",
    },
    1,
    "reference 70.0 45.0 6",
  ],
  Array [
    Object {
      "payload": Object {
        "name": "temp-1",
        "type": "thermometer",
      },
      "type": "sensor",
    },
    2,
    "thermometer temp-1",
  ],
  Array [
    Object {
      "payload": 72.4,
      "type": "entry",
    },
    3,
    "2007-04-05T22:00 72.4",
  ],
  Array [
    Object {
      "payload": 76,
      "type": "entry",
    },
    4,
    "2007-04-05T22:01 76.0",
  ],
  Array [
    Object {
      "payload": 79.1,
      "type": "entry",
    },
    5,
    "2007-04-05T22:02 79.1",
  ],
  Array [
    Object {
      "payload": 75.6,
      "type": "entry",
    },
    6,
    "2007-04-05T22:03 75.6",
  ],
  Array [
    Object {
      "payload": 71.2,
      "type": "entry",
    },
    7,
    "2007-04-05T22:04 71.2",
  ],
  Array [
    Object {
      "payload": 71.4,
      "type": "entry",
    },
    8,
    "2007-04-05T22:05 71.4",
  ],
  Array [
    Object {
      "payload": 69.2,
      "type": "entry",
    },
    9,
    "2007-04-05T22:06 69.2",
  ],
  Array [
    Object {
      "payload": 65.2,
      "type": "entry",
    },
    10,
    "2007-04-05T22:07 65.2",
  ],
  Array [
    Object {
      "payload": 62.8,
      "type": "entry",
    },
    11,
    "2007-04-05T22:08 62.8",
  ],
  Array [
    Object {
      "payload": 61.4,
      "type": "entry",
    },
    12,
    "2007-04-05T22:09 61.4",
  ],
  Array [
    Object {
      "payload": 64,
      "type": "entry",
    },
    13,
    "2007-04-05T22:10 64.0",
  ],
  Array [
    Object {
      "payload": 67.5,
      "type": "entry",
    },
    14,
    "2007-04-05T22:11 67.5",
  ],
  Array [
    Object {
      "payload": 69.4,
      "type": "entry",
    },
    15,
    "2007-04-05T22:12 69.4",
  ],
  Array [
    Object {
      "payload": Object {
        "name": "temp-2",
        "type": "thermometer",
      },
      "type": "sensor",
    },
    16,
    "thermometer temp-2",
  ],
  Array [
    Object {
      "payload": 69.5,
      "type": "entry",
    },
    17,
    "2007-04-05T22:01 69.5",
  ],
  Array [
    Object {
      "payload": 70.1,
      "type": "entry",
    },
    18,
    "2007-04-05T22:02 70.1",
  ],
  Array [
    Object {
      "payload": 71.3,
      "type": "entry",
    },
    19,
    "2007-04-05T22:03 71.3",
  ],
  Array [
    Object {
      "payload": 71.5,
      "type": "entry",
    },
    20,
    "2007-04-05T22:04 71.5",
  ],
  Array [
    Object {
      "payload": 69.8,
      "type": "entry",
    },
    21,
    "2007-04-05T22:05 69.8",
  ],
  Array [
    Object {
      "payload": Object {
        "name": "hum-1",
        "type": "humidity",
      },
      "type": "sensor",
    },
    22,
    "humidity hum-1",
  ],
  Array [
    Object {
      "payload": 45.2,
      "type": "entry",
    },
    23,
    "2007-04-05T22:04 45.2",
  ],
  Array [
    Object {
      "payload": 45.3,
      "type": "entry",
    },
    24,
    "2007-04-05T22:05 45.3",
  ],
  Array [
    Object {
      "payload": 45.1,
      "type": "entry",
    },
    25,
    "2007-04-05T22:06 45.1",
  ],
  Array [
    Object {
      "payload": Object {
        "name": "hum-2",
        "type": "humidity",
      },
      "type": "sensor",
    },
    26,
    "humidity hum-2",
  ],
  Array [
    Object {
      "payload": 44.4,
      "type": "entry",
    },
    27,
    "2007-04-05T22:04 44.4",
  ],
  Array [
    Object {
      "payload": 43.9,
      "type": "entry",
    },
    28,
    "2007-04-05T22:05 43.9",
  ],
  Array [
    Object {
      "payload": 44.9,
      "type": "entry",
    },
    29,
    "2007-04-05T22:06 44.9",
  ],
  Array [
    Object {
      "payload": 43.8,
      "type": "entry",
    },
    30,
    "2007-04-05T22:07 43.8",
  ],
  Array [
    Object {
      "payload": 42.1,
      "type": "entry",
    },
    31,
    "2007-04-05T22:08 42.1",
  ],
  Array [
    Object {
      "payload": Object {
        "name": "mon-1",
        "type": "monoxide",
      },
      "type": "sensor",
    },
    32,
    "monoxide mon-1",
  ],
  Array [
    Object {
      "payload": 5,
      "type": "entry",
    },
    33,
    "2007-04-05T22:04 5",
  ],
  Array [
    Object {
      "payload": 7,
      "type": "entry",
    },
    34,
    "2007-04-05T22:05 7",
  ],
  Array [
    Object {
      "payload": 9,
      "type": "entry",
    },
    35,
    "2007-04-05T22:06 9",
  ],
  Array [
    Object {
      "payload": Object {
        "name": "mon-2",
        "type": "monoxide",
      },
      "type": "sensor",
    },
    36,
    "monoxide mon-2",
  ],
  Array [
    Object {
      "payload": 2,
      "type": "entry",
    },
    37,
    "2007-04-05T22:04 2",
  ],
  Array [
    Object {
      "payload": 4,
      "type": "entry",
    },
    38,
    "2007-04-05T22:05 4",
  ],
  Array [
    Object {
      "payload": 10,
      "type": "entry",
    },
    39,
    "2007-04-05T22:06 10",
  ],
  Array [
    Object {
      "payload": 8,
      "type": "entry",
    },
    40,
    "2007-04-05T22:07 8",
  ],
  Array [
    Object {
      "payload": 6,
      "type": "entry",
    },
    41,
    "2007-04-05T22:08 6",
  ],
]
`;
